src/ledger/LedgerState.cpp

```diff
@@ -0,0 +1,801 @@
+}
+
+void
+LedgerStateRoot::setPreviousHeader(LedgerHeader const& header)
+{
+    mPreviousHeader = header;
+}
+
+Database&
+LedgerStateRoot::getDatabase()
+{
+    return mDb;
+}
+
+Database::EntryCache&
+LedgerStateRoot::getCache()
+{
+    // TODO(jonjove): Do we want to continue using the cache from Database?
+    // We could refactor so the cache is a member of LedgerStateRoot
+    return mDb.getEntryCache();
+}
+
+LedgerState::LedgerState(LedgerStateRoot& root)
+    : mRoot(&root), mParent(nullptr), mChild(nullptr)
+{
+    mRoot->hasChild(true);
+}
+
+LedgerState::LedgerState(LedgerState& parent)
+    : mRoot(nullptr), mParent(&parent), mChild(nullptr)
+{
+    assert(!mParent->mChild);
+    mParent->mChild = this;
+    for (auto const& state : mParent->mState)
+    {
+        state.second->invalidate();
+    }
+    mParent->mLoadBestOfferContext.clear();
+}
+
+LedgerState::~LedgerState()
+{
+    if (mParent || mRoot)
+    {
+        rollback();
+    }
+}
+
+LedgerEntryChanges
+LedgerState::getChanges()
+{
+    assert(mParent || mRoot);
+    assert(!mChild);
+
+    LedgerEntryChanges changes;
+    for (auto const& state : mState)
+    {
+        // TODO(jonjove): Should output meta if *entry == *previous?
+        auto const& entry = state.second->ignoreInvalid().entry();
+        auto const& previous = state.second->ignoreInvalid().previousEntry();
+        if (previous)
+        {
+            changes.emplace_back(LEDGER_ENTRY_STATE);
+            changes.back().state() = *previous;
+            if (entry)
+            {
+                changes.emplace_back(LEDGER_ENTRY_UPDATED);
+                changes.back().updated() = *entry;
+            }
+            else
+            {
+                changes.emplace_back(LEDGER_ENTRY_REMOVED);
+                changes.back().removed() = state.first;
+            }
+        }
+        else
+        {
+            if (entry)
+            {
+                changes.emplace_back(LEDGER_ENTRY_CREATED);
+                changes.back().created() = *entry;
+            }
+        }
+    }
+    return changes;
+}
+
+std::vector<LedgerEntry>
+LedgerState::getLiveEntries()
+{
+    assert(mRoot);
+    assert(!mChild);
+
+    std::vector<LedgerEntry> entries;
+    for (auto const& state : mState)
+    {
+        auto le = state.second->ignoreInvalid().entry();
+        if (le)
+        {
+            entries.push_back(*le);
+        }
+    }
+    return entries;
+}
+
+std::vector<LedgerKey>
+LedgerState::getDeadEntries()
+{
+    assert(mRoot);
+    assert(!mChild);
+
+    std::vector<LedgerKey> keys;
+    for (auto const& state : mState)
+    {
+        if (state.second->ignoreInvalid().entry() &&
+            !state.second->ignoreInvalid().previousEntry())
+        {
+            keys.push_back(state.first);
+        }
+    }
+    return keys;
+}
+
+void
+LedgerState::commit(std::function<void()> f)
+{
+    assert(mParent || mRoot);
+    assert(!mChild);
+
+    if (mParent)
+    {
+        assert(!f);
```

Comment:
please use more descriptive name for that

Reply:
Definitely a good point. Renamed to `onCommitToDatabase`. This was added to allow some other SQL changes to be bundled into the SQL transaction that writes `LedgerState` to database. This will be used during ledger close as some things are written, for example, to `PersistentState` within the same SQL transaction.